### PR TITLE
[WIP] Localization API

### PR DESCRIPTION
**Pull request by windy1**
[**API**](https://github.com/SpongePowered/SpongeAPI/pull/1088) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/525) | [Forge](https://github.com/SpongePowered/SpongeForge/pull/549)

This PR adds a localization API to SpongeAPI among some other breaking changes. The `provideFirst(Class)` method has been added to ServiceManager to allow for extensible services and `getServiceManager` has been added to `PluginContainer` for internal services. In the context of localization, this allows plugins to register their own `Dictionary` and have it be received with `plugin.getServiceManager().provideFirst(Dictionary.class);` without the need to know the exact implementing class.

----

**Example usage:**

```java
@Listener
public void onGameStarted(GameStartedServerEvent event) {
    DefaultPluginDictionary dict = new DefaultPluginDictionary(this, Locales.DEFAULT, configDir);
    try {
        dict.load();
    } catch (IOException e) {
        e.printStackTrace();
    }
    Sponge.getPluginManager().getServiceManager(this).setProvider(Cause.of(this), Dictionary.class, dict);
}
```

This will create a new dictionary that will first look for the dictionary file in `<configDir>/<pluginId>.dict` and then as a resource at `<pluginId>.dict`. Plugins are of course free to implement their localization through dictionaries however they want.

**Example .dict file:**

```
en_US {
    hello="Hello"
}
en_GB {
    hello="'Ello govna'"
}
```

```java
dictionary.get("hello").get(); // "Hello"
dictionary.get("hello", Locale.UK).get(); // "'Ello govna'"

// alternatively

Text.get(this, "hello").get(); // Text{"Hello"}
Text.get(this, "hello", Locale.UK).get(); // Text{"'Ello govna'"}

// localize for player's client locale

Text.get(this, "hello", player.getLocale()).get();
```

### CODE REVIEW

File: src/main/java/org/spongepowered/api/service/ServiceManager.java [hunk 1]
```diff
@@ -48,14 +50,14 @@
...
-     * @param plugin The instance of a plugin
+     * @param cause The cause of the provider change.
      * @param service The service
      * @param provider The implementation
      * @param <T> The type of service
      * @throws IllegalArgumentException Thrown if {@code plugin} is not a
      *     plugin instance
      */
-    <T> void setProvider(Object plugin, Class<T> service, T provider);
+    <T> void setProvider(Cause cause, Class<T> service, T provider);
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/service/ServiceManager.java]: Why this change? Services are supposed to be owned by a plugin

File: src/main/java/org/spongepowered/api/locale/RemoteDictionary.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+package org.spongepowered.api.locale;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Locale;
+
+/**
+ * Represents a {@link Dictionary} with a remote source.
+ */
+public interface RemoteDictionary extends Dictionary {
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/RemoteDictionary.java]: `InputStreamSource`? "Remote" isn't very specific

File: src/main/java/org/spongepowered/api/locale/AbstractDictionary.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Abstract implementation of {@link Dictionary}.
+ */
+public abstract class AbstractDictionary implements Dictionary {
+
+    protected final Object subject;
+    protected final Locale defaultLocale;
+
+    public AbstractDictionary(Object subject, Locale defaultLocale) {
+        checkNotNull(subject, "subject");
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/AbstractDictionary.java]: `this.subject = checkNotNull(subject, "subject");`

File: src/main/java/org/spongepowered/api/locale/ConfigDictionaryMulti.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    protected final Map<Locale, ConfigurationNode> localeNodes = new HashMap<>();
+
+    public ConfigDictionaryMulti(Object subject, Locale defaultLocale) {
+        super(subject, defaultLocale);
+    }
+
+    @Override
+    public ConfigurationNode load(Locale locale) throws IOException {
+        ConfigurationNode node = ConfigDictionary.super.load(locale);
+        localeNodes.put(locale, node);
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/ConfigDictionaryMulti.java]: `this.`

File: src/main/java/org/spongepowered/api/locale/ConfigDictionaryMulti.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        return node;
+    }
+
+    @Override
+    public Optional<ConfigurationNode> getNode(Locale locale) {
+        return Optional.ofNullable(localeNodes.get(locale));
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/ConfigDictionaryMulti.java]: `this.`

File: src/main/java/org/spongepowered/api/locale/ConfigDictionarySingle.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public ConfigurationNode load(Locale locale) throws IOException {
+        return this.root = ConfigDictionary.super.load(locale);
+    }
+
+    @Override
+    public Optional<ConfigurationNode> getNode(Locale locale) {
+        return Optional.ofNullable(root);
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/ConfigDictionarySingle.java]: `this.`

File: src/main/java/org/spongepowered/api/text/Text.java [hunk 7]
```diff
@@ -971,6 +975,162 @@ public static Text of(Object... objects) {
...
     /**
+     * Returns the Text value for the specified key in the specified
+     * {@link Dictionary} for the specified {@link Locale}.
+     *
+     * @param dictionary Dictionary to search
+     * @param key Key to search for
+     * @param locale Locale target
+     * @return Localized text
+     */
+    public static Text get(Dictionary dictionary, String key, Locale locale, String def) {
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/text/Text.java]: I'd personally like to keep this separate from the text class directly because the text class is just supposed to represent the things supported directly by Minecraft. Can we move this to something like a `TextDict` class?

File: src/main/java/org/spongepowered/api/locale/RemoteDictionary.java (hunk 2 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/locale/RemoteDictionary.java]: I honestly think it describes it perfectly and is less verbose than InputStreamSource. All it is is a Dictionary with a "remote" source.

File: src/main/java/org/spongepowered/api/locale/ConfigDictionarySingle.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    protected ConfigurationNode root;
+
+    public ConfigDictionarySingle(Object subject, Locale defaultLocale) {
+        super(subject, defaultLocale);
+    }
+
+    @Override
+    public ConfigurationNode load(Locale locale) throws IOException {
+        return this.root = ConfigDictionary.super.load(locale);
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/ConfigDictionarySingle.java]: Pretty sure this should be like:
```java
this.root = ConfigDictionary.super.load(locale);
return this.root;
```

also, why `ConfigDictionary.super/` instead of `super/`?

File: src/main/java/org/spongepowered/api/locale/ResourceDictionary.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Locale;
+
+/**
+ * Represents a {@link FileDictionary} that retrieves it's source from a
+ * ClassLoader resource.
+ */
+public interface ResourceDictionary extends FileDictionary {
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/ResourceDictionary.java]: `ClassLoaderResourceDictionary`? `Resource` isn't very specific imo

File: src/main/java/org/spongepowered/api/plugin/PluginContainer.java [hunk 10]
```diff
@@ -72,4 +73,11 @@ default Logger getLogger() {
      */
     Optional<Object> getInstance();
 
+    /**
+     * Returns the plugin's internal service manager.
+     *
+     * @return Internal service manager
+     */
+    ServiceManager getServiceManager();
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/plugin/PluginContainer.java]: huh?

File: src/main/java/org/spongepowered/api/locale/RemoteDictionary.java (hunk 2 above)
Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/RemoteDictionary.java]: Any why should a "remote" source be restricted to an `InputStream`?

File: src/main/java/org/spongepowered/api/service/ServiceManager.java (hunk 1 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/service/ServiceManager.java]: Services being owned by a plugin was completely arbitrary previously. The plugin is never used within the registration so there's really no need for this restriction.

Comment (kashike) [inline src/main/java/org/spongepowered/api/service/ServiceManager.java]: How am I to know which plugin owns a service provider?

File: src/main/java/org/spongepowered/api/plugin/PluginContainer.java (hunk 10 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/plugin/PluginContainer.java]: Please read the PR description. Also I would appreciate it if you could refrain from making formatting / grammar comments at this time. As I said this is still very much a WIP and the groundwork is just being laid.

File: src/main/java/org/spongepowered/api/locale/RemoteDictionary.java (hunk 2 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/locale/RemoteDictionary.java]: InputStream is not really a restriction, you can have it read a FileInputStream, StringReader, etc.

File: src/main/java/org/spongepowered/api/service/ServiceManager.java (hunk 1 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/service/ServiceManager.java]: You check the cause and see if it's an instance of a plugin.

File: src/main/java/org/spongepowered/api/locale/ConfigDictionarySingle.java (hunk 8 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/locale/ConfigDictionarySingle.java]: @kashike Default methods in interfaces must be called this way.

**Comment by kashike** [comment]
I don't see a reason for a plugin to have a `ServiceManager`, I can just see it causing confusion:
- where do I set providers, global or plugin?
- where do I get a provider?

Perhaps something else (`PluginServiceManager`, etc) to distinguish between the two?

**Comment by windy1** [comment]
@kashike The reasoning behind this is to allow for extensible services on a per-plugin basis. Consider the following.

```java
abstract class AbstractService {
}

class MyService extends AbtractService {
}

plugin.getServiceManager().setProvider(Cause.of(plugin), MyService.class, new MyService());
```

Now others can retrieve the plugin's service with

```java
plugin.getServiceManager().provideFirst(AbstractService.class);
```

**Comment by kashike** [comment]
Yeah, I'm just worried that people will start confusing the global service manager with the per-plugin one - issues arising from things not working as expected, etc.

**Comment by windy1** [comment]
@kashike The methods are well documented, sure it's a diversion from the original system, but people will get used to it.

File: src/main/java/org/spongepowered/api/text/Text.java (hunk 7 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/text/Text.java]: Hmm yeah, I see your point but I personally like this from a design standpoint and since these are static the "representation" of the Text object is the same anyways. But the location of these methods will definitely be worth discussing down the line a bit.

File: src/main/java/org/spongepowered/api/service/SimpleServiceManager.java [hunk 11]
```diff
@@ -97,6 +73,20 @@ public SimpleServiceManager(PluginManager pluginManager) {
 
     @SuppressWarnings("unchecked")
     @Override
+    public <T> Optional<T> provideFirst(Class<T> service) {
+        checkNotNull(service, "service");
+        for (Map.Entry<Class<?>, ProviderRegistration<?>> entry : this.providers.entrySet()) {
+            Class<?> s = entry.getKey();
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/service/SimpleServiceManager.java]: one char var? :person_frowning:

**Comment by zml2008** [comment]
There should be a ResourceBundleDictionary -- since ResourceBundle is basically the standard way to handle localization in java and handles a lot of the file matching logic already.

File: src/main/java/org/spongepowered/api/locale/RemoteDictionary.java (hunk 2 above)
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/locale/RemoteDictionary.java]: Using a local or bundled file has nothing to do with "remote". IMO the current name does not express properly what this class does.
If `InputStreamDictionary` is too long, then truncate it to `StreamDictionary`

File: src/main/java/org/spongepowered/api/locale/ResourceDictionary.java (hunk 9 above)
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/locale/ResourceDictionary.java]: IMO a `Resource` isn't a `File`. So this shouldn't extend `FileDictionary`. You can see that its something else if you look at `getSource`. Also the `Path` part is missused because it is used as a string and does not work at all outside this class.
If you need a layer that both share use `URI`s.

File: src/main/java/org/spongepowered/api/service/ServiceManager.java (hunk 1 above)
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/service/ServiceManager.java]: What else do you want to specify using Cause?
IMO a plugin just provides a Service. There is no cause like `Player x entered y`.
And well the cause `plugin is installed` does not count.

File: src/main/java/org/spongepowered/api/text/Text.java (hunk 7 above)
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/text/Text.java]: Maybe we can move this to `Dictionary` as `getAsText(...) : Optional<Text>`.

File: src/main/java/org/spongepowered/api/locale/ConfigDictionarySingle.java (hunk 8 above)
Comment (Zidane) [inline src/main/java/org/spongepowered/api/locale/ConfigDictionarySingle.java]: @kashike 

That line assignment would happen before the return so this is perfectly valid java.

**Comment by windy1** [comment]
@zml2008 Added ResourceBundleDictionary interface. ConfigDictionary now extends this and also wraps config nodes in a resource bundle.

File: src/main/java/org/spongepowered/api/locale/AbstractRemoteDictionary.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+     *
+     * @return Source resolver
+     */
+    public SourceResolver getResolver() {
+        return this.resolver;
+    }
+
+    @Override
+    public InputStream getSource(Locale locale) throws Exception {
+        return this.resolver.resolve(locale).orElseThrow(() -> new IllegalStateException("Could not resolve source for locale " + locale + "."));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/locale/AbstractRemoteDictionary.java]: could uh, use `checkState(false, "")` instead of a lambda ;)

**Comment by windy1** [comment]
Updated description with example usage.

File: src/main/java/org/spongepowered/api/locale/ConfigDictionarySingle.java (hunk 8 above)
Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/ConfigDictionarySingle.java]: I'm aware it is valid Java, I'm just not sure if it's something that the code style allows - I haven't looked at every line in Sponge, but I haven't seen any other uses of this myself.

File: src/main/java/org/spongepowered/api/locale/ResourceBundleDictionary.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+     * @param locale Locale to set
+     * @param bundle Bundle to use for Locale
+     */
+    void setBundle(Locale locale, B bundle);
+
+    @Override
+    default Optional<String> get(String key, Locale locale) {
+        checkNotNull(key, "key");
+        checkNotNull(locale, "locale");
+        return Optional.of(getBundle(locale).getString(key));
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/ResourceBundleDictionary.java]: This will throw, not return null.
https://docs.oracle.com/javase/8/docs/api/java/util/ResourceBundle.html#getString-java.lang.String-

Comment (windy1) [inline src/main/java/org/spongepowered/api/locale/ResourceBundleDictionary.java]: This will throw what? `MissingResourceException`? If that's what you're referring to than yes, I'm aware, and it's the desired result.

Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/ResourceBundleDictionary.java]: Yup - catch and return `empty` is what I meant

Comment (windy1) [inline src/main/java/org/spongepowered/api/locale/ResourceBundleDictionary.java]: Thanks, fixed now.

File: src/main/java/org/spongepowered/api/locale/RemoteDictionary.java (hunk 2 above)
Comment (Mumfrey) [inline src/main/java/org/spongepowered/api/locale/RemoteDictionary.java]: @kashike, @ST-DDT honestly you're just nitpicking here, `RemoteDictionary` is a perfectly cromulent name.

File: src/main/java/org/spongepowered/api/service/ServiceManager.java (hunk 1 above)
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/service/ServiceManager.java]: I agree - I don't think a cause makes sense here. What would be in the cause, other than a plugin container?

Comment (windy1) [inline src/main/java/org/spongepowered/api/service/ServiceManager.java]: Yeah I'm likely going to change this back actually. Originally I was going to pass a `Game` object for the game dictionary but I ended up just passing the dummy plugin instead.

File: src/main/java/org/spongepowered/api/locale/Dictionary.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+ * given string key and return a localized result.
+ */
+public interface Dictionary {
+
+    /**
+     * Returns the "subject" for this dictionary.
+     *
+     * @return Subject of dictionary
+     */
+    Object getSubject();
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/locale/Dictionary.java]: Could ever be something other than a plugin? If so, `Dictionary` should probably have a generic parameter for the subject. If not, could this be a `PluginContainer` instead?

Comment (kashike) [inline src/main/java/org/spongepowered/api/locale/Dictionary.java]: I'd personally prefer a generic, no need to limit it to plugins.

File: src/main/java/org/spongepowered/api/locale/DefaultPluginDictionary.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+public class DefaultPluginDictionary extends DefaultDictionary {
+
+    /**
+     * File extension used in this implementation.
+     */
+    public static final String FILE_EXTENSION = ".dict";
+
+    public DefaultPluginDictionary(Object plugin, Locale defaultLocale, Path dir) {
+        super(plugin, defaultLocale, Paths.get(dir.toString(), Sponge.getPluginManager().getId(plugin) + FILE_EXTENSION));
+    }
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/locale/DefaultPluginDictionary.java]: You should use `dir.resolve(Sponge.getPluginManager()...) here`

File: src/main/java/org/spongepowered/api/plugin/PluginManager.java [hunk 16]
```diff
@@ -76,4 +90,53 @@
      */
     boolean isLoaded(String id);
 
+    /**
+     * Gets the id of the {@link Plugin}.
+     *
+     * @return The id
+     */
+    default String getId(Object plugin) {
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/plugin/PluginManager.java]: I don't think this is a good idea, especially because #1081 adds much more things here. Why not call `find(plugin).getId()` directly? I'd doubt this would make a big difference.

Comment (windy1) [inline src/main/java/org/spongepowered/api/plugin/PluginManager.java]: Purely for convenience. IMO the less verbose we can make these kind of things the better. Similar concept to the `Sponge` class really.

Comment (Minecrell) [inline src/main/java/org/spongepowered/api/plugin/PluginManager.java]: There always has to be a limit for these kind of things, and if it results in a lot of duplicated code while not applicable in too many use cases this is exactly the limit IMO. In most cases plugins will want to check several of these methods anyway, and in that case it is much more efficient to call the `find` method once and then call the methods directly.

Comment (kashike) [inline src/main/java/org/spongepowered/api/plugin/PluginManager.java]: I agree with @Minecrell here

File: src/main/java/org/spongepowered/api/plugin/PluginManager.java [hunk 17]
```diff
@@ -45,6 +48,17 @@
     Optional<PluginContainer> fromInstance(Object instance);
 
     /**
+     * Finds the {@link PluginContainer} for the specified plugin instance.
+     *
+     * @param instance Plugin instance
+     * @return Plugin container of instance
+     * @throws PluginNotFoundException if the instance could not be found
+     */
+    default PluginContainer find(Object instance) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/plugin/PluginManager.java]: What's the point in this method exactly, that `fromInstance` doesn't already provide, other than throwing an exception?

Comment (windy1) [inline src/main/java/org/spongepowered/api/plugin/PluginManager.java]: Removes boilerplate in plugins. If I want to obtain an instance of my own plugin container one would want an exception to be thrown if it's not found.

Comment (Minecrell) [inline src/main/java/org/spongepowered/api/plugin/PluginManager.java]: If you want your own plugin container you would usually add this in your main class:

```java
@Inject PluginContainer myContainer;
```

Comment (windy1) [inline src/main/java/org/spongepowered/api/plugin/PluginManager.java]: Fair enough, I still think it's worth keeping this method though.

Comment (Minecrell) [inline src/main/java/org/spongepowered/api/plugin/PluginManager.java]: `fromInstance(Object)` returns an optional. Normally you could just call `get` if you assume it is present and it will also throw an exception otherwise.

File: src/main/java/org/spongepowered/api/text/Text.java [hunk 18]
```diff
@@ -933,6 +936,37 @@ public static Text of(Object... objects) {
         return builder.build();
     }
 
+    public static Optional<Text> get(Dictionary dictionary, String key, Locale locale) {
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/text/Text.java]: This should be still in an extra class IMO.

Comment (Zidane) [inline src/main/java/org/spongepowered/api/text/Text.java]: I disagree, I go to Text to get what I want from Text.

File: src/main/java/org/spongepowered/api/plugin/PluginManager.java (hunk 17 above)
Comment (Zidane) [inline src/main/java/org/spongepowered/api/plugin/PluginManager.java]: I agree with Minecrell...the get would throw an exception and you can listen for that exception instead.

File: src/main/java/org/spongepowered/api/text/Text.java (hunk 18 above)
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/text/Text.java]: IMO it will just confusion if there is for example `Text.of(String)` and `Text.get(String)` even though both methods are **completely** different.

Comment (windy1) [inline src/main/java/org/spongepowered/api/text/Text.java]: Ideally, developers will pretty much only use `Text.get(String)` to receive their messages. For the most part constant strings should not be used for text creation.

Comment (Minecrell) [inline src/main/java/org/spongepowered/api/text/Text.java]: You'll always need a simple string from time to time..... Sending messages is only one area where Text is used, it's used in a lot of other places as well.

File: src/main/java/org/spongepowered/api/locale/AbstractRemoteDictionary.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+     *
+     * @return Source resolver
+     */
+    public SourceResolver getResolver() {
+        return this.resolver;
+    }
+
+    @Override
+    public InputStream getSource(Locale locale) throws Exception {
+        checkNotNull(locale, "locale");
```
Comment (octo-byte) [inline src/main/java/org/spongepowered/api/locale/AbstractRemoteDictionary.java]: Magic string "locale" is used across multiple files of the super AbstractDictionary. Might be a good idea to move it into a constant on the super

File: src/main/java/org/spongepowered/api/locale/RemoteDictionary.java [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+        /**
+         * Resolves the {@link InputStream} source for the given
+         * {@link Locale}.
+         *
+         * @param locale Locale to resolve source for
+         * @return Optional InputStream. Empty if unresolved.
+         * @throws Exception
+         */
+        public Optional<InputStream> resolve(Locale locale) throws Exception {
+            checkNotNull(locale, "locale");
```
Comment (octo-byte) [inline src/main/java/org/spongepowered/api/locale/RemoteDictionary.java]: Again, magic this magic string is kind of scattered all over the place. If you can find somewhere convenient to put it, that'd be cool.

Comment (wizjany) [inline src/main/java/org/spongepowered/api/locale/RemoteDictionary.java]: if you're talking about the second argument to the checkNotNull method, that's an error message that gets spit out if locale is null, not some sort of magic string. error messages are hardcoded all over the place.

**Comment by octo-byte** [comment]
I'm quickly skimming over this and just wondering - language files can get pretty big, pretty quickly (see Android for example).
It might be a good idea to keep one locale per .dict file, under a directory (eg. `<configDir>/<pluginId>/<locale>.dict`)

File: src/main/java/org/spongepowered/api/locale/AbstractRemoteDictionary.java (hunk 19 above)
Comment (evilmidget38) [inline src/main/java/org/spongepowered/api/locale/AbstractRemoteDictionary.java]: This seems to be a standard practice across the Sponge project.  The String "locale" has no functional purpose and exists purely for the purpose of providing convenient stack traces, so I would argue against it being placed in a static final field.  Furthermore, "locale" is referencing the name of the parameter.  It is only coincidental that it is utilized across multiple files.

Comment (octo-byte) [inline src/main/java/org/spongepowered/api/locale/AbstractRemoteDictionary.java]: Ah okay. Thanks for clarifying.

**Comment by windy1** [comment]
@octo-byte that is certainly going to be an option when this pr is complete. See `MultiSourceConfigDictionary`.

**Comment by VcSaJen** [comment]
I'm sorry if it's obvious from the code, but…
How it's gonna work with per-user localization (client-locale-dependent) in: 
- commands argument keys
- command descriptions
- permission description
- TextMessageException
- Other instances where Text isn't outputted directly by plugin, but stored and outputted by Sponge.

**Comment by windy1** [comment]
@VcSaJen Per-player localization will look something like this:

```java
player.sendMessage(Text.get("message.key", player.getLocale()));
```

**Comment by VcSaJen** [comment]
@windy1 I'm talking about localizing `CommandSpec.builder().description(...` and `.extendedDescription`.

**Comment by kashike** [comment]
Continuing in #1206
